src/server/frontend_xwayland/xwayland_surface_observer.cpp

```diff
@@ -113,6 +120,22 @@ void mf::XWaylandSurfaceObserver::input_consumed(ms::Surface const*, MirEvent co
 {
     if (mir_event_get_type(event) == mir_event_type_input)
     {
+        auto const input_event = mir_event_get_input_event(event);
+        if (mir_input_event_get_type(input_event) == mir_input_event_type_key)
+        {
+            // Ignore key repeat events
+            if (mir_keyboard_event_action(mir_input_event_get_keyboard_event(input_event)) ==
+                mir_keyboard_action_repeat)
+                return;
+
+            if (!has_focus)
+            {
+                log_warning("**** Sending key to surface even though it was not explicitly given keyboard focus");
+            }
+            // If we're sending keys we implicitly have input focus
+            focussed(true);
```

Comment:
I don't like all this. It duplicates logic in `wl_keyboard.cpp`, and we never should be hitting this at all.